projects/nlp/clustering_runner.py

```diff
@@ -0,0 +1,363 @@
+
+  colorSequenceBucket = []
+  for docId in documentCategoryMap:
+    buckets = documentCategoryMap[docId]
+    counts = categoryCounts[buckets]
+    maxBucketIndex = numpy.argmax(counts)
+    maxBucket = buckets[maxBucketIndex]
+    colorSequenceBucket.append(maxBucket)
+
+  plt.figure()
+  plt.subplot(121, aspect="equal")
+  plt.title("Bucket labels (%s)" % (args.modelName,))
+  plt.xlabel("PC 2")
+  plt.ylabel("PC 1")
+  plt.scatter(projectedData[:,1], projectedData[:,0], c=colorSequenceBucket)
+  
+  colorSequenceClusters = numpy.zeros(len(colorSequenceBucket))
+  clusterId = 0
+  for dataIndices in protos:
+    colorSequenceClusters[[d for d in dataIndices if d != -1]] = clusterId
+    clusterId += 1
+  
+  plt.subplot(122, aspect="equal")
+  plt.title("Clusters (%s)" % (args.modelName,))
+  plt.xlabel("PC 2")
+  plt.ylabel("PC 1")
+  plt.scatter(projectedData[:,1], projectedData[:,0], c=colorSequenceClusters)
+  plt.savefig(os.path.join(SAVE_PATH, "scatter.png"))
+
+  plt.figure()
+  plt.plot(s[:250])
+  plt.xlabel("Singular value #")
+  plt.ylabel("Singular value")
+  plt.savefig(os.path.join(SAVE_PATH, "singular_values.png"))
+
+
+def createBucketClusterPlot(args, bucketCounts):
+  bucketCounts += 0
+  bucketDist = bucketCounts / bucketCounts.sum(1, keepdims=True)
+  
+  plt.pcolor(bucketDist, cmap=plt.cm.Blues)
+  # plt.bar(range(len(labelRefs)), bucketDist.T)
+  plt.xlabel("Bucket")
+  plt.ylabel("Cluster")
+  plt.title("%d clusters using model %s" % (bucketCounts.shape[0], args.modelName))
+  
+  for rowIx in xrange(bucketCounts.shape[0]):
+    for colIx in xrange(bucketCounts.shape[1]):
+      if bucketCounts[rowIx, colIx] != 0:
+        plt.annotate(str(int(bucketCounts[rowIx, colIx])), xy=(colIx+0.2, rowIx+0.4))
+
+  plt.savefig(os.path.join(SAVE_PATH, "bucket_cluster_matrix.png"))
+  
+
+
+def knnTest(protos, knn):
+  print "Running KNN Test"
+  testPassed = True
+  protoSets = [set(protoList)-set([-1]) for protoList in protos]
+
+  for cluster in protoSets:
+    if len(cluster) == 1:
+      continue
+      
+    for index in cluster:
+      if index != -1:
+        # print index
+        testPattern = knn.getPattern(index)
+        allDistances = knn._getDistances(testPattern)
+        closest = allDistances.argmin()
+        # print closest
+        if closest not in cluster:
+          testPassed = False
+  if testPassed:
+    print "KNN test passed"
+  else:
+    print "KNN test failed"
+
+
+def trainModel(args, model, trainingData, labelRefs):
+  """
+  Train the given model on trainingData. Return the trained model instance.
+  """
+  print
+  print "=======================Training model on sample text================"
+  for docId, doc in enumerate(trainingData):
+    document = doc[0]
+    labels = doc[1]
+    print
+    print "Document=", document, "label=",labelRefs[doc[1][0]], "id=",docId
+    model.trainDocument(document, labels, docId)
+
+  return model
+
+
+def readData(args):
```

Comment:
I moved this method into csv_helper as readDataAndReshuffle()  Could you use that version instead?